### PR TITLE
Allow null timestamp values

### PR DESCRIPTION
**Pull request by charlesfries**

### CODE REVIEW

File: addon/transforms/timestamp.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ export default class Timestamp extends Transform {
   }
 
   serialize(deserialized) {
-    if (typeOf(deserialized) !== 'date') {
+    if (typeOf(deserialized) !== 'date' && typeOf(deserialized) !== 'null') {
```
Comment (mikkopaderes) [inline addon/transforms/timestamp.js:16]: Should this be `null` instead of `'null'`? `'null'` would end up being a string, right?

**Review by mikkopaderes — CHANGES_REQUESTED** [review]
Thanks for this. I understand the use-case and yeah, we should support it. But we also need to update the docs and unit tests to cover this new feature. Would you mind changing it as well? Should be a quick one 😄

File: addon/transforms/timestamp.js (hunk 1 above)
Comment (mikkopaderes) [reply addon/transforms/timestamp.js:16]: Nevermind, I didn't notice you're doing a `typeOf` here. Making it `deserialized !== null` should work as well and is more straightforward.

**Comment by charlesfries** [comment]
Yes, I'll get on it later tonight 👍

**Comment by charlesfries** [comment]
Finally got around to this. Seems like there are some conflicts however. But I updated tests and wrote some more documentation 👌🏻

**Comment by mikkopaderes** [comment]
> 
> 
> Finally got around to this. Seems like there are some conflicts however. But I updated tests and wrote some more documentation 👌🏻

Yes, thanks. The conflict is because the files you're editing are now on TypeScript.

**Comment by charlesfries** [comment]
Ah I see. Just pushed another commit to update the files to TS.

Also lmk if the documentation is too verbose, I tried to show all three use cases for the transform

File: docs/transforms.md [hunk 2]
```diff
@@ -10,11 +10,56 @@ import Model, { attr } from '@ember-data/model';
 export default class PostModel extends Model {
   @attr title;
   @attr('timestamp') createdOn;
+  @attr('timestamp', { defaultValue: null }) modifiedOn;
+  @attr('timestamp', { defaultValue: null }) publishOn;
```
Comment (mikkopaderes) [inline docs/transforms.md:14]: These are docs that I feel that we don't need. It should be covered by Ember Data itself.

File: docs/transforms.md [hunk 3]
```diff
@@ -10,11 +10,56 @@ import Model, { attr } from '@ember-data/model';
...
+  "posts": {
+    "post_a": {
+      "createdOn": "November 10, 2020 at 12:00:00 AM UTC-8",
+      "modifiedOn": "November 10, 2020 at 12:00:05 AM UTC-8",
+      "publishOn": "December 25, 2020 at 12:00:00 AM UTC-8",
+      "title": "My Post (Edited)"
+    }
+  }
+}
+```
```
Comment (mikkopaderes) [inline docs/transforms.md:62]: Just like you mentioned, I feel that the docs makes things too verbose. Your [line 18](https://github.com/mikkopaderes/ember-cloud-firestore-adapter/pull/133/files#diff-b7cdfd453c5a6d190f2c990a55cf8439815dc9ac1bb1e0a94b362110370d2dd6R18) I believe should be enough to cover things in a simple manner.

**Review by mikkopaderes — CHANGES_REQUESTED** [review]
There's an existing test that's failing because it was using the `null` value and expecting a Firestore timestamp. Would it be okay for you to change that as well? :smile:

**Comment by mikkopaderes** [comment]
Thinking about this feature more, I think I may have jumped a bit to early on agreeing that we should support this.

Right now, I can't imagine any generic use-case for this and it may just add confusion to the `Timestamp` transform on why it's allowed to be `null`. If anyone needs this, I'd recommend to create your own custom transform for your application called `nullable-timestamp`.

Thank you for taking time on working on this PR and I'm sorry as well if this seems like lost time on your side. 😢

**Pull request closed without merging**